### ISSUE TITLE
Path to GraphiQL v2

### ISSUE THREAD

**Issue by timsuchanek** [graphql/graphiql#2328]
We have [ambitious plans](https://github.com/graphql/graphiql/discussions/2216) for the future of GraphiQL and want to go there step by step.

This is the path I suggest to move forward:

## 1. Implement [new design](https://www.figma.com/file/p8f8MaG7ZRnuCc4lVv7w6t/GraphiQL-2.0?node-id=0%3A1) with existing components
- [x] https://github.com/graphql/graphiql/issues/2325
## 2. Modernize components step by step
As the new design has a slightly different layout, in Step 1, the components can already be modernized as needed.
If necessary, we'll create a separate issue for tracking work that's left.

## 3. Introduce monaco
- [ ] https://github.com/graphql/graphiql/issues/2326

## 4. Implement the new sidebar.
The new sidebar is quite ambitious, that's why we want to work on it last.
- [ ] https://github.com/graphql/graphiql/issues/2327
<img width="502" alt="image" src="https://user-images.githubusercontent.com/1094804/164035768-1fbc91f3-5115-4764-9f08-9cc662aba818.png">

## Additional tasks
While working on this, we want to revisit and consolidate the boards @acao created:
- [ ] https://github.com/orgs/graphql/projects/1/views/1?query=is%3Aopen+sort%3Aupdated-desc
- [ ] https://github.com/orgs/graphql/projects/1/views/4?query=is%3Aopen+sort%3Aupdated-desc
- [ ] https://github.com/orgs/graphql/projects/1/views/5?query=is%3Aopen+sort%3Aupdated-desc
- [ ] https://github.com/orgs/graphql/projects/1/views/7?query=is%3Aopen+sort%3Aupdated-desc

Some of the issues will be fixed by our tasks above, some are additional improvements, bug fixes, and features.

We'll have @thomasheyenbrock from GraphCDN focus on steps 1-4 in the coming weeks.
Anyone who's interested in collaborating, please let us know and we can organize code pairing sessions!

We also commit to 1h of issue triaging per week, as some of them grew a bit old. We start May 2nd at 4 pm UTC. Let me know if you want to join. Calendar link:

https://calendar.google.com/calendar/u/0/embed?src=linuxfoundation.org_ik79t9uuj2p32i3r203dgv5mo8@group.calendar.google.com

You can either answer here or [join the graphiql Discord channel.](https://discord.gg/cXKXpdMyTt)

**Comment by acao** [comment]
@timsuchanek the boards are just filters based on label, you can remove them readily but then you'll have a huuuge list if that's ok!

**Comment by timsuchanek** [comment]
Awesome, thanks for the context, we'll have a look into those tasks once we tackle any `monaco-graphql` related things. I'll also ping you again before working on it just to make sure we don't miss anything, as you have most of the context.

**Comment by acao** [comment]
Oh perhaps you missed the discussion on discord where I will be working on monaco editor adoption concurrently!

**Comment by n1ru4l** [comment]
How will this path be represented in releases?
When I think that each step could be an individual release step 1 would already be a huge breaking change if shipped 🤔 .
The following order seems to make more sense to me:

1. Modernize components step by step (new minor)
2. Introduce Monaco (new minor as there are no APIs for directly hooking into codemirror AFAIK)
**Entering breaking change land**
4. Implement [new design](https://www.figma.com/file/p8f8MaG7ZRnuCc4lVv7w6t/GraphiQL-2.0?node-id=0%3A1) with existing components (major - alongside with this release it might then also make sense to revisit GraphiQL props and adjust the API)
5. Implement the new sidebar (new minor)

**Comment by acao** [comment]
this makes sense to me! this is almost the exact same roadmap we took in 2020 and it worked out great. the only problem is we didn't release incrementally.

we can most of the components [converted to hooks](https://github.com/graphql/graphiql/blob/main/packages/graphiql-2-rfc-context/src/components/DocExplorer/SchemaDoc.tsx) and [context](https://github.com/graphql/graphiql/tree/main/packages/graphiql-2-rfc-context/src/api/providers) two years ago, so that should be a great starting point for anyone who wants to embark on this again! Just ignore the monaco specifics, and localistaion

**Comment by acao** [comment]
monaco itself will also incur a ton of breaking changes, such as entirely different keybindings, different customTheme prop, etc, so it might warrant a 2.0 itself, and the redesign a 3.0

**Comment by thomasheyenbrock** [comment]
Just to clarify @n1ru4l @acao: Would you release the redesign with a new major version even when there wouldn't be changes to the API (exported methods, component props, etc.)? Or do you just think it's not possible to avoid breaking changes there given the new design?

**Comment by acao** [comment]
@thomasheyenbrock adding `monaco-graphql` *would* necessitate breaking changes to the props API

**Comment by thomasheyenbrock** [comment]
I agree, Monaco is for sure a breaking change. In my understanding, introducing Monaco and reworking the UI are two separate tasks that don't have to be done in one release. Would you rather say that these two have to go hand in hand?

**Comment by acao** [comment]
@thomasheyenbrock I don't think they need to go hand in hand. one can happen without the other, in whichever order.

**Comment by thomasheyenbrock** [comment]
I agree, Monaco is for sure a breaking change. I agree, introducing Monaco and reworking the UI are two separate tasks that don't have to be done in one release 👍 Also, I don't have anything against releasing the redesign as v3 since the way we'll change the UI can be considered breaking even though there might not be actual API changes if we do Monaco in a separate release.

**Comment by acao** [comment]
yes, the API changes caused by the monaco release will be to the `editorTheme` prop, and a few other `GraphiQL` props. very minor breaking changes. from a release-notes-as PR/FAQ method, we can propose the monaco release will work like this:

> congrats! there are only a few breaking changes in this release. the greatest breaking change will involve learning the new keybindings. _here is a document_ that explains the new monaco keybindings. if you're used to vscode, you'll find a lot of familiarity :)

**Comment by timsuchanek** [comment]
Thanks, everyone for bringing in your perspective!
It seems like we agree on the breakdown of tasks and have different ideas for the potential order.
I'm curious to hear, what makes you think that the design is a breaking change @n1ru4l @acao?

The reason I added that to the front, was, that for GraphiQL@1 users I want us to provide value as quickly as possible.

I'm generally definitely open to doing a major release for the design as well, just want to understand what you see breaking in there.

We seem to have a consensus, that Monaco is a breaking change, due to the reasons @acao alluded to.

I also think that the people working on this can see which order fits best. What matters is that we just keep the breakdown in releases, which we already agree on, which is great.

**Comment by timsuchanek** [comment]
@patrick91 just added you to the event!

**Comment by jonathanawesome** [comment]
I've just started digging into the work that @thomasheyenbrock has been doing to move components into `@graphiql/react`, as Step 1 above, and the breaking change concerns from @n1ru4l have become clearer to me. 

Leaving aside the differences in the props/API surface for now, which I haven't fully explored, I think most of the _breaking_ UI changes have to do with the children pattern that currently looks like [this](https://github.com/graphql/graphiql/tree/main/packages/graphiql#children-this-pattern-will-be-dropped-in-200):

> <GraphiQL.Logo>: Replace the GraphiQL logo with your own.
> 
> <GraphiQL.Toolbar>: Add a custom toolbar above GraphiQL. If not provided, a default toolbar may contain common operations. Pass the empty <GraphiQL.Toolbar /> if an empty toolbar is desired.
> 
> <GraphiQL.Button>: Add a button to the toolbar above GraphiQL.
> 
> <GraphiQL.Menu>: Add a dropdown menu to the toolbar above GraphiQL.
> 
> <GraphiQL.MenuItem>: Items for a menu.
> <GraphiQL.Select>: Add a select list to the toolbar above GraphiQL.
> 
> <GraphiQL.SelectOption>: Options for a select list.
> <GraphiQL.Group>: Add a group of associated controls to the toolbar above GraphiQL. Expects children to be <GraphiQL.Button>, <GraphiQL.Menu>, or <GraphiQL.Select>.
> 
> <GraphiQL.Footer>: Add a custom footer below GraphiQL Results.


#### Logo
The "logo" has been relocated to an area that, potentially, could have an impact on the layout. For example,Yoga adds an icon and extra copy to the default "GraphiQL" copy. 

<img width="653" alt="Screen Shot 2022-07-29 at 6 04 12 PM" src="https://user-images.githubusercontent.com/963528/181857368-d6249f1a-b060-4b33-9296-1b349a275cca.png">

<img width="498" alt="Screen Shot 2022-07-29 at 6 10 07 PM" src="https://user-images.githubusercontent.com/963528/181859045-f83d6be9-5dd3-4d34-a4d3-b6c32dc9c608.png">

I imagine there are other implementations that do something similar, so we'll need to either constrain that prop in 2.0 or offer an alternative branding option. Either way, it's currently accepting arbitrary content and seems like a breaking change to me.

#### Toolbar / Menu / MenuItem / Select / SelectOption / Group
This is the big one...the toolbar of old does not exist in the new design. Instead, it's functionality is split between the vertically-arranged actions within the operations editor (run operation, prettify, copy, etc) and the sidebar/plugin navigation (history, explorer, etc). 

Yoga:
<img width="471" alt="Screen Shot 2022-07-29 at 6 04 59 PM" src="https://user-images.githubusercontent.com/963528/181857586-3d55a884-6ded-4235-b077-9a8c93c2fb4b.png">

Github's [implementation](https://docs.github.com/en/graphql/overview/explorer) which they refer to as "Explorer" and uses the OneGraph Explorer:
<img width="425" alt="Screen Shot 2022-07-29 at 5 49 15 PM" src="https://user-images.githubusercontent.com/963528/181855121-d9a885a0-58d7-4cd8-9853-457ca018e5dc.png">

This is easily a breaking change, and becomes gnarlier when you consider how many implementations are using OneGraph's Explorer. We can surface a new API for Operation Actions, but I don't know if it's realistic to ship a v2 without a plan for addressing OneGraph Explorer usage. This would leave a lot of implementations just plain stuck, as their primary interface _is_ OneGraph's Explorer.

#### Footer
This is a weird one. I don't know why it's called "Footer", must be a legacy thing, and the only implementation I'm aware of is @benjie's [postgraphiql](https://github.com/graphile/postgraphile/blob/4164d437e57a30ff00ecbb4675943a0924c54ec9/postgraphiql/src/components/PostGraphiQL.js#L830) (I'm sure there are others...if you're reading, please comment!). 

We do have some designs that _might_ map to the current architecture:
<img width="346" alt="Screen Shot 2022-07-29 at 7 09 24 PM" src="https://user-images.githubusercontent.com/963528/181862211-68787822-f7d7-4ace-832c-b7e63dbd01bb.png">

But it's still a breaking change because the new design doesn't have a singular "footer" space for arbitrary content.

So...all of that to say that I'm convinced that the new design, as wonderful and beautiful and necessary as it is, is a breaking change to existing implementations. We have more thinking to do around what is included in a 2.0 release.

**Comment by acao** [comment]
@jonathanawesome we had an issue pinned for years that announces the deprecation of these interfaces for 2.x, it’s still in the readme i think

Either way it would be good to provide a migration path! These are widely used and the baseline of 2.0.0 we promised is that each oss framework could do some basic branding customizations at least - color theme, maybe typefaces, and logo at least. 

Perhaps the other regions aren’t as important now that we’re shipping the editor with new ways to customize. As long as we can allow them to add their own  sidebar components that use the hooks, and be able to add commands, they should have most of the flexibility they need. 

If they need a more customized layout, they can “eject” and copy over the few components that make up graphiql and compose their own editor

**Comment by benjie** [comment]
For context, PostGraphiQL's footer serves two purposes:

1. a single line detailing helpful links to the docs, examples, support, and of course sponsorship ;)
2. a large expansion containing the SQL that was executed from the GraphQL query/the SQL query plan/etc

Point 1 would be nice to have somewhere. Point 2 would be much better served by the alternative views that the new design allows for _however_ it is pretty important (to me at least) to be able to see the query, the result, AND the SQL queries all at the same time. So I wouldn't like to just _replace_ the result with the SQL queries/etc, that would make it concretely less useful to me.

Basically what I'm saying is that point 2 shouldn't be in the footer, it's only there because there was nowhere else suitable that we could put it in the older GraphiQL. Plugins should help.

Point 1 could easily go under the entire GraphiQL interface as a separate component I suppose; much like the footer on a website. It may not need to be factored into the 2.x design.

**Comment by thomasheyenbrock** [comment]
I know that the child pattern has been announced to be dropped in v2, but I'm not sure we should as it's widely used.
- The logo can still be customized using `GraphiQL.Logo`, this should allow the GraphQL Yoga use-case where a custom icon is added next to the "GraphiQL" text.
- The toolbar can be customized in two ways:
  - Additional content can be added using the `toolbar` prop, leaving the existing buttons (prettify, copy, and merging fragments) in place
  - The default toolbar can be replaced by passing `GraphiQL.Toolbar` as child.
  
  The former would allow GraphQL Yoga and PostGraphile to add buttons to the toolbar.
- Regarding the footer, in the current state of our v2 efforts it still exists. In my opinion the "response tools" we have in the design are too much for v2. They are basically part of a fully thought through plugin API, and I don't think this is feasible in v2. We wanted to ship as quickly as possible, that's why I think v2 should only take care of introducing the new design, _excluding_ the doc explorer and the complete plugin structure.

**Comment by acao** [comment]
> but I'm not sure we should as it's widely used

excellent point, and [you are so correct](https://cs.github.com/?scopeName=All+repos&scope=&q=GraphiQL.Logo)! 

and it's not impossible to implement or type either! I'm sure you've found some ways to improve it @thomasheyenbrock I haven't even looked. it's not even an uncommon pattern. And it's convenient. And with the hooks, we can do so much more than before! Toolbar buttons and Footer will will be even more powerful than before with `@graphiql/react`!

Everyone is making great points. Should this move to a discussion about the future of the child prop pattern? We can copy and paste some of these comments to start their own threads honestly! haha

**Comment by jonathanawesome** [comment]
I'm not a huge fan of the children pattern going forward, as the new design allows for _more_ customization and I think exposing these options as typed props is a tighter experience. For those users that find a typed interface too tight, they'd likely be comfortable plucking the components out of `@graphiql/react` and building something of their own.

> * The logo can still be customized using `GraphiQL.Logo`, this should allow the GraphQL Yoga use-case where a custom icon is added next to the "GraphiQL" text.

I'm suggesting that allowing arbitrary content, which is what `GraphiQL.Logo` is, is _not_ the correct solution considering the new design. Not only is horizontal space in that Tab row at a premium, but it doesn't seem like the right place for a mark/lockup/logo.

I'd like to see the GraphiQL type _always_ be visible in that spot and serve as a link to the repo/documentation. 

Any sort of branding and/or arbitrary content, a la @benjie's use case, might be better served by using a Pane Plugin. This doesn't need to be available via the full-blown plugin interface, we could offer a much simpler interface for v2 that allows for something like this:

<img width="453" alt="Screen Shot 2022-07-30 at 9 36 10 AM" src="https://user-images.githubusercontent.com/963528/181919165-52eecbe1-9361-459e-af44-7a8aea658dbb.png">

👆This is also a great solution for removing the wall of welcome/how-to text that's currently in the operations editor on initial load.
<img width="503" alt="Screen Shot 2022-07-30 at 9 44 39 AM" src="https://user-images.githubusercontent.com/963528/181919500-a41ebb2f-d6c1-4708-9c7b-376b3e363008.png">


> * The toolbar can be customized in two ways:
>   
>   * Additional content can be added using the `toolbar` prop, leaving the existing buttons (prettify, copy, and merging fragments) in place
>   * The default toolbar can be replaced by passing `GraphiQL.Toolbar` as child.
>   
>   The former would allow GraphQL Yoga and PostGraphile to add buttons to the toolbar.

The toolbar has been re-contextualized, so any buttons that are being added (by either of the two methods) that aren't operation actions will be very much out of place. [Here's what PostGraphiQl does](https://github.com/graphile/postgraphile/blob/4164d437e57a30ff00ecbb4675943a0924c54ec9/postgraphiql/src/components/PostGraphiQL.js#L789). 

> * Regarding the footer, in the current state of our v2 efforts it still exists. In my opinion the "response tools" we have in the design are too much for v2. They are basically part of a fully thought through plugin API, and I don't think this is feasible in v2. We wanted to ship as quickly as possible, that's why I think v2 should only take care of introducing the new design, _excluding_ the doc explorer and the complete plugin structure.

Totally agree re: response tools and shipping a well-formed plugin structure for v2. My concern with `GraphiQL.Footer` being used as-is in v2 is that if it doesn't work for some implementations (it's arbitrary content, who knows how many variations exist) we'll end up scrambling for a solution just to get those folks into v2.

**Comment by thomasheyenbrock** [comment]
@jonathanawesome great points you bring up here!

Regarding the logo: I like the idea of putting it to the top left, but there are other subtleties to consider here like you already mentioned, like reusing the plugin pane on the left to have a dedicated view that shows custom content. Instead of figuring this out for v2, I'd propose to keep the current way where a user can switch out the "logo" next to the tabs on the top right as it's a very flexible solution. (Yes we do allow arbitrary content, but we can't protect GraphiQL users from messing up the design in all cases anyways.)

Just to clarify around the footer: Implementations that use `GraphiQL.Footer` today should not need to change anything about it in order to upgrade to v2 since the placement and the layout is currently still the same.

I like the idea of replacing the child pattern with typed props eventually, again I just think that's too much for v2. I believe we have enough breaking changes for that major version, and adding more will just make it harder for folks to upgrade. It might also be possible to introduce these props in future minors, officially deprecate the child pattern and then remove it in v3 or a future major update.

TL;DR I see most of this out of scope for v2, but would love for us to further discuss it with more focus by opening dedicated issues.

**Comment by SimenB** [comment]
Now v2 is out (:tada:) should this be closed and/or unpinned?

**Comment by Sytten** [comment]
I am also wondering if the refactor of the various components is done since the underlining issues have not been closed. We are looking to build another implementation for vue people and we would like to know of we can use a couple building blocks that would already be present.

**Comment by thomasheyenbrock** [comment]
Thanks for the callouts @SimenB and @Sytten! We should definitely do some housekeeping, close out some historic issues that have been resolved with v2 and consolidate the various roadmap-issues. I'll try to spend some time over the coming days to clean this up.
